firmware/src/Adafruit_BME680_rpi.cpp

```diff
@@ -0,0 +1,624 @@
+/*!
+ * @file Adafruit_BME680.cpp
+ *
+ * @mainpage Adafruit BME680 temperature, humidity, barometric pressure and gas
+ * sensor driver
+ *
+ * @section intro_sec Introduction
+ *
+ * This is the documentation for Adafruit's BME680 driver for the
+ * Arduino platform.  It is designed specifically to work with the
+ * Adafruit BME680 breakout: https://www.adafruit.com/products/3660
+ *
+ * These sensors use I2C to communicate, 2 pins (SCL+SDA) are required
+ * to interface with the breakout.
+ *
+ * Adafruit invests time and resources providing this open source code,
+ * please support Adafruit and open-source hardware by purchasing
+ * products from Adafruit!
+ *
+ * @section author Author
+ *
+ * Written by Ladyada for Adafruit Industries.
+ *
+ * @section license License
+ *
+ * BSD license, all text here must be included in any redistribution.
+ *
+ */
+
+#include "Adafruit_BME680_rpi.h"
+// #include "Arduino.h"
+
+//#define BME680_DEBUG
+
+/** Our hardware interface functions **/
+// static int8_t i2c_read(uint8_t reg_addr, uint8_t *reg_data, uint32_t len,
+//                        void *interface);
+// static int8_t i2c_write(uint8_t reg_addr, const uint8_t *reg_data, uint32_t len,
+//                         void *interface);
+static int8_t spi_read(uint8_t reg_addr, uint8_t *reg_data, uint32_t len,
+                       void *interface);
+static int8_t spi_write(uint8_t reg_addr, const uint8_t *reg_data, uint32_t len,
+                        void *interface);
+static void delay_usec(uint32_t us, void *intf_ptr);
+
+unsigned long millis();
+
+void delay(uint32_t ms);
+
+
+
+/*!
+ *  @brief  Reads 8 bit values over SPI
+ */
+static int8_t spi_read(uint8_t reg_addr, uint8_t *reg_data, uint32_t len,
+                       void *intf_ptr) {
+  MySPI *_dev = (MySPI *)intf_ptr;
+
+  reg_addr |= 0x80;
+
+  _dev->read(reg_addr,reg_data,len);
+
+  return 0;
+
+  // if (!_dev->write_then_read(&reg_addr, 1, reg_data, len, 0x0)) {
+  //   return -1;
+  // }
+
+  // return 0;
+}
+
+/*!
+ *  @brief  Writes 8 bit values over SPI
+ */
+static int8_t spi_write(uint8_t reg_addr, const uint8_t *reg_data, uint32_t len,
+                        void *intf_ptr) {
+  MySPI *_dev = (MySPI *)intf_ptr;
+
+  _dev->write(reg_addr,reg_data,len);
+
+  // if (!_dev->write((uint8_t *)reg_data, len, &reg_addr, 1)) {
+  //   return -1;
+  // }
+
+  return 0;
+}
+
+// Shim functions so copied Adafruit code can use arduino names
```

Comment:
Hmm, to make this file more standalone and truer to the original without edits, I'd suggest pulling out all native Arduino-style into their own \*.cpp and \*.h file called *Arduino.h*. 

In there, I would also include a Serial shim class.
Here's [an example](https://github.com/AllenNeuralDynamics/harp.pico.multistage/blob/main/firmware/inc/Arduino.h) from another project.

Reply:
Okay, so I moved things into Arduino.h/cpp, but I didn't flesh out the Serial shim class as it's going to need overloads for all the types that Serial.print accepts. I made an issue (#6) so we can remember to get to it later though.